xdsl/parser.py

```diff
@@ -1,1335 +1,1924 @@
+        info = self.input.get_lines_containing(self)
+        if info is None:
+            return "Unknown location of span {}. Error: ".format(self, msg)
+        lines, offset_of_first_line, line_no = info
+        # offset relative to the first line:
+        offset = self.start - offset_of_first_line
+        remaining_len = max(self.len, 1)
+        capture = StringIO()
+        print("{}:{}:{}".format(self.input.name, line_no, offset),
+              file=capture)
+        for line in lines:
+            print(line, file=capture)
+            if remaining_len < 0:
+                continue
+            len_on_this_line = min(remaining_len, len(line) - offset)
+            remaining_len -= len_on_this_line
+            print("{}{}".format(" " * offset, "^" * max(len_on_this_line, 1)),
+                  file=capture)
+            if msg is not None:
+                print("{}{}".format(" " * offset, msg), file=capture)
+                msg = None
+            offset = 0
+        if msg is not None:
+            print(msg, file=capture)
+        return capture.getvalue()
+
+    def __repr__(self):
+        return "{}[{}:{}](text='{}')".format(self.__class__.__name__,
+                                             self.start, self.end, self.text)
+
+
+@dataclass(frozen=True, repr=False)
+class StringLiteral(Span):
 
-    def __str__(self):
-        return f"{self.line}:{self.column}"
+    def __post_init__(self):
+        if len(self) < 2 or self.text[0] != '"' or self.text[-1] != '"':
+            raise ParseError(self, "Invalid string literal!")
 
-    def next_char_pos(self, n: int = 1) -> Position | None:
-        """Return the position of the next character in the string."""
-        if self.idx >= len(self.file) - n:
+    @classmethod
+    def from_span(cls, span: Span | None) -> StringLiteral | None:
+        if span is None:
             return None
-        new_idx = self.idx
-        new_line = self.line
-        new_column = self.column
-        while n > 0:
-            if self.file[new_idx] == '\n':
-                new_line += 1
-                new_column = 1
-            else:
-                new_column += 1
-            new_idx += 1
-            n -= 1
-        assert new_idx < len(self.file)
-        return Position(self.file, new_idx, new_line, new_column)
-
-    def get_char(self) -> str:
-        """Return the character at the current position."""
-        assert self.idx < len(self.file)
-        return self.file[self.idx]
-
-    def get_current_line(self) -> str:
-        """Return the current line."""
-        assert self.idx < len(self.file)
-        start_idx = self.idx - self.column + 1
-        end_idx = self.idx
-        while self.file[end_idx] != '\n':
-            end_idx += 1
-        return self.file[start_idx:end_idx]
+        return cls(span.start, span.end, span.input)
 
+    @property
+    def string_contents(self):
+        # TODO: is this a hack-job?
+        return ast.literal_eval(self.text)
```

Comment:
as far as I'm aware, this is a "safe" api [see docs](https://docs.python.org/3/library/ast.html#ast.literal_eval):

>  The string or node provided may only consist of the following Python literal structures: strings, bytes, numbers, tuples, lists, dicts, sets, booleans, None and Ellipsis.

I didn't want to re-invent the wheel here, so instead of writing my own string unescape function, I took one that the python stdlib provides.


I have to admit though, that writing this didn't feel super good, that's why the `TODO` is there. If anyone has a better plan, please let me know.

Reply:
What is this supposed to do actually?
Is it just supposed to return a string?

Reply:
I think that the parser should transform string literals like `"test\ntest"` into strings that contain an actual newline character. And instead of writing a routine that replaces all escaped special characters myself, I thought I should use a library function for that. And that is exactly what `ast.literal_eval` does.

@math-fehr I think that your old parser would not do the special character replacement thing. I would really like to understand why that is, is it an MLIR thing? Because IMO a parser should absolutely replace escape sequences in string literals.